# DSCI 525 - Web and Cloud Computing

## Milestone 1: Tackling big data on your laptop

### Content:

1. [Downloading the data](#1)
2. [Combining data CSVs](#2)
3. [Load the combined CSV to memory and perform a simple EDA](#3)
4. [Perform a simple EDA in R](#4)

In [19]:
# Most of the code is adapted from lecture notes

In [1]:
import re
import os
import glob
import zipfile
import requests
from urllib.request import urlretrieve
import json
import pandas as pd
from memory_profiler import memory_usage
import numpy as np
import dask.dataframe as dd

import pyarrow.dataset as ds
import rpy2_arrow.pyarrow_rarrow as pyra
import pyarrow.feather as feather
import pyarrow.parquet as pq

In [2]:
# %load_ext rpy2.ipython
%load_ext memory_profiler

<a id='1'/>

### 1. Downloading the data

1. [Download the data from figshare to your local computer using the figshare API (you can make use of requests library).](#download_data)
2. [Extract the zip file, again programmatically, similar to how we did it in class.](#extract_zip)



In [3]:
# add metadata
article_id = 14096681  # this is the unique identifier of the article on figshare
url = f"https://api.figshare.com/v2/articles/{article_id}"
headers = {"Content-Type": "application/json"}
output_directory = "data/"

In [4]:
# define 
response = requests.request("GET", url, headers=headers)
data = json.loads(response.text)  # this contains all the articles data
files = data["files"]             # this is just the data about the files, which is what we want

In [5]:
# create directory if doesn't exist
os.makedirs(output_directory, exist_ok=True)

<a id='download_data'/>

#### 1. Download the data from figshare

In [59]:
%%time
%%memit

# download data using the API
files_to_dl = ["data.zip"]
for file in files:
    if file["name"] in files_to_dl:       
        urlretrieve(file["download_url"], output_directory + file["name"])

<a id='extract_zip'/>

#### 2.Extract the zip file

In [15]:
%%time
%%memit      

# unzip the files
with zipfile.ZipFile(os.path.join(output_directory, "data.zip"), 'r') as f:
    f.extractall(output_directory)


peak memory: 295.71 MiB, increment: 0.14 MiB
CPU times: user 17.4 s, sys: 2.33 s, total: 19.7 s
Wall time: 22.5 s


<a id='2'/>

### 2. Combining data CSVs

1. [Use one of the following options to combine data CSVs into a single CSV.](#combine_csv)
    - Pandas (we chose this)
    - DASK
2. When combining the csv files make sure to add extra column called "model" that identifies the model.
3. Compare run times and memory usages of these options on different machines within your team, and summarize your observations in your milestone notebook.

    - Run times for combining CSV files for all team members (in alphabetical order):
        - **Fatime**
             - peak memory: 9823.55 MiB, increment: 9558.20 MiB
             - CPU times: user 1min 3s, sys: 12.5 s, total: 1min 16s
             - Wall time: 1min 25s
        - **Jingjing**
             - peak memory: 431.68 MiB, increment: 0.25 MiB
             - CPU times: user 5min 40s, sys: 23.9 s, total: 5min 46s
             - Wall time: 6min 27s  

In [6]:
# all csv files
files = glob.glob(output_directory + '*.csv')

# remove csvs that are formatted differently
files = [file for file in files if "observed" not in file]

<a id='combine_csv'/>   

#### 1. Combine CSV files into one big CSV file

In [7]:
usecols = ['time', 'lat_min', 'lat_max', 'lon_min', 'lon_max', 'rain (mm/day)']

In [8]:
%%time
%%memit

# create a dict of dataframes
df = {file.split('_daily')[0].split('/')[1]: pd.read_csv(file, usecols=usecols) for file in files}

# concatenate to one dataframe
df = pd.concat(df, names=['model']).reset_index(level=1, drop=True)

peak memory: 10519.62 MiB, increment: 10255.38 MiB
CPU times: user 1min 1s, sys: 10.2 s, total: 1min 11s
Wall time: 1min 14s


In [9]:
df.head()

time    lat_min    lat_max   lon_min  \
model                                                                  
MPI-ESM-1-2-HAM  1889-01-01 12:00:00 -35.439867 -33.574619  141.5625   
MPI-ESM-1-2-HAM  1889-01-02 12:00:00 -35.439867 -33.574619  141.5625   
MPI-ESM-1-2-HAM  1889-01-03 12:00:00 -35.439867 -33.574619  141.5625   
MPI-ESM-1-2-HAM  1889-01-04 12:00:00 -35.439867 -33.574619  141.5625   
MPI-ESM-1-2-HAM  1889-01-05 12:00:00 -35.439867 -33.574619  141.5625   

                  lon_max  rain (mm/day)  
model                                     
MPI-ESM-1-2-HAM  143.4375   4.244226e-13  
MPI-ESM-1-2-HAM  143.4375   4.217326e-13  
MPI-ESM-1-2-HAM  143.4375   4.498125e-13  
MPI-ESM-1-2-HAM  143.4375   4.251282e-13  
MPI-ESM-1-2-HAM  143.4375   4.270161e-13

In [10]:
df.tail()

time    lat_min   lat_max  lon_min  lon_max  \
model                                                                     
SAM0-UNICON  2014-12-27 12:00:00 -30.157068 -29.21466  153.125  154.375   
SAM0-UNICON  2014-12-28 12:00:00 -30.157068 -29.21466  153.125  154.375   
SAM0-UNICON  2014-12-29 12:00:00 -30.157068 -29.21466  153.125  154.375   
SAM0-UNICON  2014-12-30 12:00:00 -30.157068 -29.21466  153.125  154.375   
SAM0-UNICON  2014-12-31 12:00:00 -30.157068 -29.21466  153.125  154.375   

             rain (mm/day)  
model                       
SAM0-UNICON       6.689683  
SAM0-UNICON       7.862555  
SAM0-UNICON      10.005026  
SAM0-UNICON       8.541592  
SAM0-UNICON      68.117489

In [11]:
df.to_csv('combined.csv', chunksize=1000)

<a id='3'/>

### 3. Load the combined CSV to memory and perform a simple EDA

##### [1. Investigate at least two of the following approaches to reduce memory usage while performing the EDA (e.g., value_counts).](#investigate_approaches)
- [Changing dtype of your data](#change_dtype)
- [Load just columns what we want](#load_cols)
- [Loading in chunks](#load_chunks)
- [Dask](#dask)

##### [2. Discuss your observations.](#discuss1)

<a id='investigate_approaches'/>

#### 1. Investigate at least two of the following approaches to reduce memory usage while performing the EDA (e.g., value_counts)

---
#### Establish baseline relevant info about the combined data

In [11]:
df.head()

time    lat_min    lat_max   lon_min  \
model                                                                  
MPI-ESM-1-2-HAM  1889-01-01 12:00:00 -35.439867 -33.574619  141.5625   
MPI-ESM-1-2-HAM  1889-01-02 12:00:00 -35.439867 -33.574619  141.5625   
MPI-ESM-1-2-HAM  1889-01-03 12:00:00 -35.439867 -33.574619  141.5625   
MPI-ESM-1-2-HAM  1889-01-04 12:00:00 -35.439867 -33.574619  141.5625   
MPI-ESM-1-2-HAM  1889-01-05 12:00:00 -35.439867 -33.574619  141.5625   

                  lon_max  rain (mm/day)  
model                                     
MPI-ESM-1-2-HAM  143.4375   4.244226e-13  
MPI-ESM-1-2-HAM  143.4375   4.217326e-13  
MPI-ESM-1-2-HAM  143.4375   4.498125e-13  
MPI-ESM-1-2-HAM  143.4375   4.251282e-13  
MPI-ESM-1-2-HAM  143.4375   4.270161e-13

In [12]:
# check current dtype and other info
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 66151029 entries, MPI-ESM-1-2-HAM to SAM0-UNICON
Data columns (total 6 columns):
 #   Column         Dtype  
---  ------         -----  
 0   time           object 
 1   lat_min        float64
 2   lat_max        float64
 3   lon_min        float64
 4   lon_max        float64
 5   rain (mm/day)  float64
dtypes: float64(5), object(1)
memory usage: 3.5+ GB


In [30]:
%%time
%%memit

# simple EDA for baseline data
df.describe()

peak memory: 11169.43 MiB, increment: 5932.62 MiB
CPU times: user 14.1 s, sys: 9.47 s, total: 23.6 s
Wall time: 26.8 s


In [13]:
print(f"Memory usage of combined data with float64: {df.memory_usage().sum() / 1e6:.2f} MB")

Memory usage of combined data with float64: 3704.46 MB


----
<a id='#change_dtype'></a>

##### Change dtype of the data

In [14]:
# change the dtype of data
col_dtype_spec = {
    "model": "str",
    "time": "str",
    "lat_min": np.float32,
    "lat_max": np.float32,
    "lon_min": np.float32,
    "lon_max": np.float32,
    "rain (mm/day)": np.float32
}

In [15]:
%%time
%%memit

df1 = pd.read_csv('combined.csv', dtype=col_dtype_spec, parse_dates=["time"])
df1_eda = df1["model"].value_counts()
df1_eda

peak memory: 7254.16 MiB, increment: 2023.05 MiB
CPU times: user 1min 16s, sys: 7.07 s, total: 1min 23s
Wall time: 1min 27s


In [19]:
df1.head()

model                time    lat_min    lat_max   lon_min  \
0  MPI-ESM-1-2-HAM 1889-01-01 12:00:00 -35.439865 -33.574619  141.5625   
1  MPI-ESM-1-2-HAM 1889-01-02 12:00:00 -35.439865 -33.574619  141.5625   
2  MPI-ESM-1-2-HAM 1889-01-03 12:00:00 -35.439865 -33.574619  141.5625   
3  MPI-ESM-1-2-HAM 1889-01-04 12:00:00 -35.439865 -33.574619  141.5625   
4  MPI-ESM-1-2-HAM 1889-01-05 12:00:00 -35.439865 -33.574619  141.5625   

    lon_max  rain (mm/day)  
0  143.4375   4.244226e-13  
1  143.4375   4.217326e-13  
2  143.4375   4.498125e-13  
3  143.4375   4.251282e-13  
4  143.4375   4.270161e-13

In [20]:
# sanity check that dtypes has been changes
df1.dtypes

model                    object
time             datetime64[ns]
lat_min                 float32
lat_max                 float32
lon_min                 float32
lon_max                 float32
rain (mm/day)           float32
dtype: object

In [21]:
%%time
%%memit

df1.describe()

peak memory: 6485.61 MiB, increment: 2018.79 MiB
CPU times: user 9.33 s, sys: 2.87 s, total: 12.2 s
Wall time: 12.7 s


In [27]:
print(f"Memory usage of combined data with float32: {df1.memory_usage().sum() / 1e6:.2f} MB")

Memory usage of combined data with float32: 2381.44 MB


---

<a id='#load_cols'/>

##### Load just columns what we want

In [23]:
usecols = ["model", "time", "rain (mm/day)"]

In [24]:
%%time
%%memit

df2 = pd.read_csv('combined.csv', dtype=col_dtype_spec, parse_dates=["time"], usecols=usecols)
df2_eda = df2["model"].value_counts()
df2_eda

peak memory: 7624.25 MiB, increment: 3157.37 MiB
CPU times: user 1min 4s, sys: 6.78 s, total: 1min 11s
Wall time: 1min 14s


In [25]:
df2.head()

model                time  rain (mm/day)
0  MPI-ESM-1-2-HAM 1889-01-01 12:00:00   4.244226e-13
1  MPI-ESM-1-2-HAM 1889-01-02 12:00:00   4.217326e-13
2  MPI-ESM-1-2-HAM 1889-01-03 12:00:00   4.498125e-13
3  MPI-ESM-1-2-HAM 1889-01-04 12:00:00   4.251282e-13
4  MPI-ESM-1-2-HAM 1889-01-05 12:00:00   4.270161e-13

In [26]:
df2.dtypes

model                    object
time             datetime64[ns]
rain (mm/day)           float32
dtype: object

In [28]:
%%time
%%memit

df2.describe()

peak memory: 6246.20 MiB, increment: 1009.25 MiB
CPU times: user 2.77 s, sys: 776 ms, total: 3.55 s
Wall time: 4.53 s


In [29]:
print(f"Memory usage of combined data with only columns that we want and changed dtypes: {df2.memory_usage().sum() / 1e6:.2f} MB")

Memory usage of combined data with only columns that we want and changed dtypes: 1323.02 MB


---

<a id='load_chunks'/>

##### Loading in chunks

In [10]:
%%time
%%memit

counts = pd.Series(dtype=int)

for chunk in pd.read_csv("combined.csv", chunksize=10_000_000):
    counts = counts.add(chunk["model"].value_counts(), fill_value=0)
counts.astype(int).sort_values(ascending = False)

peak memory: 2437.45 MiB, increment: 2173.19 MiB
CPU times: user 1min 8s, sys: 5.47 s, total: 1min 13s
Wall time: 1min 22s


---

<a id='dask'></a>

##### Dask

In [13]:
%%time
%%memit

ddf = dd.read_csv('combined.csv')
ddf["model"].value_counts().compute()

peak memory: 4667.44 MiB, increment: 704.34 MiB
CPU times: user 1min 32s, sys: 15.1 s, total: 1min 47s
Wall time: 56.8 s


In [14]:
ddf.head()

model                 time    lat_min    lat_max   lon_min  \
0  MPI-ESM-1-2-HAM  1889-01-01 12:00:00 -35.439867 -33.574619  141.5625   
1  MPI-ESM-1-2-HAM  1889-01-02 12:00:00 -35.439867 -33.574619  141.5625   
2  MPI-ESM-1-2-HAM  1889-01-03 12:00:00 -35.439867 -33.574619  141.5625   
3  MPI-ESM-1-2-HAM  1889-01-04 12:00:00 -35.439867 -33.574619  141.5625   
4  MPI-ESM-1-2-HAM  1889-01-05 12:00:00 -35.439867 -33.574619  141.5625   

    lon_max  rain (mm/day)  
0  143.4375   4.244226e-13  
1  143.4375   4.217326e-13  
2  143.4375   4.498125e-13  
3  143.4375   4.251282e-13  
4  143.4375   4.270161e-13

In [15]:
from dask.diagnostics import ProgressBar

In [18]:
with ProgressBar():
    print(ddf.isnull().sum().compute())

[########################################] | 100% Completed | 52.2s
model                  0
time                   0
lat_min          3219300
lat_max                0
lon_min          3219300
lon_max                0
rain (mm/day)    3219300
dtype: int64


---
<a id='discuss1'></a>

#### 2. Discuss your observations

We have investigated all four approaches to perform a simple EDA in Python and below are the observations based on the results that we've got when we performed EDA of the combined csv file. 
- Changing dtype of your data
> We changed the type of our data such that numeric columns that were of type 'float64' are changed to 'float32' and as a result we can see that the memory usage has been reduced substantially, i.e. from that of the baseline 3704.46 MB to 2381.44 MB. Also, we saw a huge reduction in numbers for all peak memory, Wall time, and CPU times when we used `.describe()` function to perform a simple EDA in the data with float32 compared to data with float64. 
- Load just columns what we want
> We loaded just the columns that we want to perform an EDA to, that is 'model', 'time', and 'rain (mm/day)' and the time it took pandas to load the csv file was comparable to the previous apprach. The memory usage is lower which is to be expected as we are loading less columns here in this approach compared to the previous approach. In addition, the memory usage is substantially lower here compared to the data with float62.
- Loading in chunks
> In this approach we are loading the combined csv data in chunks of 10000000 records and thus serially performing EDA of model value counts. The values of Wall time and CPU time in this approach are similar to those of the previous approaches but the peak memory is way lower in this approach.
- Dask
> In this approach we used Dask which performes EDA in a more optimised way by internally loading chunks of data. Compared to the previous approach where we loaded data in chunks, the CPU time is higher here as well as 

We investigated two approaches,first using chunks and second using dask.
The first approach loads the combined csv file in chuncks of 10_000_000 and serially performs EDA of counting models
The second approach using dask internally loads in chunks and performs eda parallely in an optimized way.
The memory usage in both the first and second approach is similar.
The CPU time is higher in the dask approach as compared to first approach. Also there is higher difference between CPU time and Wall time in dask than first approach
The wall time of second approach using dask is better than the first approach

<a id='4'/>


### 4. Perform a simple EDA in R

1. [Pick an approach to transfer the dataframe from python to R.](#transfer_data)
    - [Parquet file](#transfer_parquet)
    - [Feather file](#transfer_feather)
    - [Pandas exchange](#transfer-pandas)
    - [Arrow exchange](#transfer-arrow)
2. [Discuss why you chose this approach over others.](#4-discussion)

##### Pandas format

In [4]:
%%time
%%memit
df = pd.read_csv("combined.csv", nrows=1_000_000)

peak memory: 412.40 MiB, increment: 136.86 MiB
Wall time: 1.7 s


##### Arrow format

In [8]:
%%time
%%memit
dataset = ds.dataset("combined.csv", format="csv")
table = dataset.to_table()

peak memory: 4552.66 MiB, increment: 4181.79 MiB
Wall time: 36.5 s


##### Feather format

In [9]:
%%time
%%memit
feather.write_feather(table, 'combined.feather')

peak memory: 4566.12 MiB, increment: 13.46 MiB
Wall time: 4.93 s


##### Parquet format

In [10]:
%%time
%%memit
pq.write_table(table, 'combined.parquet')

peak memory: 4708.31 MiB, increment: 142.86 MiB
Wall time: 16.5 s


In [11]:
%%sh
# csv data
du -sh combined.csv

# feather data
du -sh combined.feather

# parquet data
du -sh combined.parquet

6.0G	combined.csv
1.1G	combined.feather
542M	combined.parquet


<a id='transfer_data'/>

#### 1. Transfer the dataframe from python to R and perform EDA

In [13]:
%%R
library(dplyr)
library(arrow)

<a id='transfer-pandas'/>

##### 1. Pandas

In [14]:
%%time
%%R -i df

start_time <- Sys.time()
print(class(df))
result <- df %>% collect() %>% summary()
end_time <- Sys.time()
print(result)
print(end_time - start_time)

[1] "data.frame"
     time              lat_min          lat_max          lon_min     
 Length:1000000     Min.   :-36.25   Min.   :-35.00   Min.   :140.6  
 Class :character   1st Qu.:-36.25   1st Qu.:-35.00   1st Qu.:142.5  
 Mode  :character   Median :-35.00   Median :-33.75   Median :146.2  
                    Mean   :-34.92   Mean   :-33.67   Mean   :146.1  
                    3rd Qu.:-33.75   3rd Qu.:-32.50   3rd Qu.:150.0  
                    Max.   :-32.50   Max.   :-31.25   Max.   :151.9  
    lon_max      rain (mm/day)         model          
 Min.   :142.5   Min.   :  0.0000   Length:1000000    
 1st Qu.:144.4   1st Qu.:  0.0082   Class :character  
 Median :148.1   Median :  0.1190   Mode  :character  
 Mean   :147.9   Mean   :  1.7961                     
 3rd Qu.:151.9   3rd Qu.:  0.8148                     
 Max.   :153.8   Max.   :432.9395                     
Time difference of 0.2850361 secs
Wall time: 46.2 s


<a id='transfer-arrow'/>

##### 2. Arrow

In [15]:
r_table = pyra.converter.py2rpy(table)

6052
rarrow.ChunkedArray: 0.0380098819732666
6052
rarrow.ChunkedArray: 0.028034687042236328
6052
rarrow.ChunkedArray: 0.03199338912963867
6052
rarrow.ChunkedArray: 0.048981428146362305
6052
rarrow.ChunkedArray: 0.0460052490234375
6052
rarrow.ChunkedArray: 0.029996156692504883
6052
rarrow.ChunkedArray: 0.030958175659179688


In [16]:
%%time
%%R -i r_table

start_time <- Sys.time()
print(class(r_table))
result <- r_table %>% collect() %>% summary()
print(class(r_table %>% collect()))
end_time <- Sys.time()
print(result)
print(end_time - start_time)

[1] "Table"       "ArrowObject" "R6"         
[1] "tbl_df"     "tbl"        "data.frame"
      time                        lat_min           lat_max      
 Min.   :1889-01-01 08:06:00   Min.   :-36       Min.   :-36.00  
 1st Qu.:1920-07-02 20:00:00   1st Qu.:-35       1st Qu.:-33.66  
 Median :1952-01-01 20:00:00   Median :-33       Median :-32.04  
 Mean   :1952-01-02 00:32:09   Mean   :-33       Mean   :-31.98  
 3rd Qu.:1983-07-02 20:00:00   3rd Qu.:-31       3rd Qu.:-30.16  
 Max.   :2014-12-31 20:00:00   Max.   :-30       Max.   :-27.91  
                               NA's   :3219300                   
    lon_min           lon_max      rain (mm/day)        model          
 Min.   :141       Min.   :141.2   Min.   :  0       Length:62467843   
 1st Qu.:143       1st Qu.:145.0   1st Qu.:  0       Class :character  
 Median :147       Median :148.1   Median :  0       Mode  :character  
 Mean   :147       Mean   :148.2   Mean   :  2                         
 3rd Qu.:150       3rd 

<a id='transfer_feather'/>

##### 3. Feather format

In [ ]:
# %%time
# %%R
# start_time <- Sys.time()
# r_table <- arrow::read_feather("combined.feather")
# print(class(r_table))
# result <- r_table %>% summary()
# end_time <- Sys.time()
# print(result)
# print(end_time - start_time)

<a id='transfer_parquet'/>

##### 4. Parquet format

In [5]:
%%time
%%R

start_time <- Sys.time()
r_table <- arrow::read_parquet("combined.parquet")
print(class(r_table))
library(dplyr)
result <- r_table %>% summary()
end_time <- Sys.time()
print(result)
print(end_time - start_time)

[1] "tbl_df"     "tbl"        "data.frame"


R[write to console]: 
Attaching package: 'dplyr'


R[write to console]: The following objects are masked from 'package:stats':

    filter, lag


R[write to console]: The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union




      time                        lat_min           lat_max      
 Min.   :1889-01-01 08:06:00   Min.   :-36       Min.   :-36.00  
 1st Qu.:1920-07-02 20:00:00   1st Qu.:-35       1st Qu.:-33.66  
 Median :1952-01-01 20:00:00   Median :-33       Median :-32.04  
 Mean   :1952-01-02 00:32:09   Mean   :-33       Mean   :-31.98  
 3rd Qu.:1983-07-02 20:00:00   3rd Qu.:-31       3rd Qu.:-30.16  
 Max.   :2014-12-31 20:00:00   Max.   :-30       Max.   :-27.91  
                               NA's   :3219300                   
    lon_min           lon_max      rain (mm/day)        model          
 Min.   :141       Min.   :141.2   Min.   :  0       Length:62467843   
 1st Qu.:143       1st Qu.:145.0   1st Qu.:  0       Class :character  
 Median :147       Median :148.1   Median :  0       Mode  :character  
 Mean   :147       Mean   :148.2   Mean   :  2                         
 3rd Qu.:150       3rd Qu.:151.3   3rd Qu.:  1                         
 Max.   :154       Max.   :155.6   Max. 

<a id='4-discussion'/>

#### 2. Discussion

From above results, our group would choose parquet format to transfer the dataframe from python to R. First, the file size of our combined data is 542M with parquet format, which is much smaller than the file sizes with csv (6.0G) and feather (1.1G) format. Second, the wall time for parquet format is shorter than the wall time for feather format. The wall time for csv is a little bit shorter than the wall time for parquet. However, consider both the file size and speed, it would be better and reasonable to choose parquet format to transfer the dataframe from python to R.

##### More simpe EDA with Parquet format

In [6]:
%%time
%%R
r_table <- arrow::read_parquet("combined.parquet")
class(r_table)

[1] "tbl_df"     "tbl"        "data.frame"
Wall time: 9.31 s


In [7]:
%%time
%%R
head(r_table)

# A tibble: 6 x 7
  time                lat_min lat_max lon_min lon_max `rain (mm/day)` model     
  <dttm>                <dbl>   <dbl>   <dbl>   <dbl>           <dbl> <chr>     
1 1889-01-01 20:06:00   -36.2     -35    141.    142.        3.29e-13 "data\\AC~
2 1889-01-02 20:06:00   -36.2     -35    141.    142.        0.       "data\\AC~
3 1889-01-03 20:06:00   -36.2     -35    141.    142.        0.       "data\\AC~
4 1889-01-04 20:06:00   -36.2     -35    141.    142.        0.       "data\\AC~
5 1889-01-05 20:06:00   -36.2     -35    141.    142.        1.05e- 2 "data\\AC~
6 1889-01-06 20:06:00   -36.2     -35    141.    142.        3.29e- 2 "data\\AC~
Wall time: 3.17 s


In [8]:
%%time
%%R
tail(r_table)

# A tibble: 6 x 7
  time                lat_min lat_max lon_min lon_max `rain (mm/day)` model     
  <dttm>                <dbl>   <dbl>   <dbl>   <dbl>           <dbl> <chr>     
1 2014-12-26 20:00:00   -30.2   -29.2    153.    154.           0.435 "data\\Ta~
2 2014-12-27 20:00:00   -30.2   -29.2    153.    154.           0.554 "data\\Ta~
3 2014-12-28 20:00:00   -30.2   -29.2    153.    154.           7.03  "data\\Ta~
4 2014-12-29 20:00:00   -30.2   -29.2    153.    154.           0.235 "data\\Ta~
5 2014-12-30 20:00:00   -30.2   -29.2    153.    154.           2.10  "data\\Ta~
6 2014-12-31 20:00:00   -30.2   -29.2    153.    154.           0.548 "data\\Ta~
Wall time: 501 ms


### Reflection

- Challenges with downloading the data
    - It takes really long time to download the large size of data because of the internet connection. I fail to download the data with my own VPN. I successfully download the data by using UBC VPN
    
- Challenges with performing a simple EDA in R
    - It takes really long time to transfer dataframe from python to R with feather format. Like more than an hour. I try to restart PC and Jupyter Lab but doesn't work. I comment the code cell for feather format and get the conclusion that it takes longer time than the other formats.